ui/src/app.tsx

```diff
@@ -17,25 +22,65 @@ import Overview from "./overview";
 enableMapSet();
 
 type AppProps = {
-  underlayNames: string[];
   entityName: string;
 };
 
 export default function App(props: AppProps) {
+  const dispatch = useAppDispatch();
+  const underlaysApi = useContext(UnderlaysApiContext);
+  const entitiesApi = useContext(EntitiesApiContext);
+
+  const underlaysState = useAsyncWithApi(
+    useCallback(() => {
+      return underlaysApi.listUnderlays({}).then((res) => {
+        if (!res?.underlays || res.underlays.length == 0) {
+          throw "No underlays are configured.";
+        }
+
+        Promise.all(
+          res.underlays.map((u) => {
+            if (!u.name) {
+              throw "Unnamed underlay.";
+            }
+            return entitiesApi.listEntities({ underlayName: u.name });
+          })
+        ).then((entitiesResList) => {
+          dispatch(
+            setUnderlays(
+              entitiesResList.map((entitiesRes, i) => {
+                const name = res.underlays?.[i]?.name;
+                if (!name) {
+                  throw "Unnamed underlay.";
+                }
+                if (!entitiesRes.entities) {
```

Comment:
Should this check length as well?

Reply:
I don't think it needs to. For our current schemas we'd always expect there to be entities but it doesn't seem like an error in general.